linkerd/tracing/src/access_log.rs

```diff
@@ -0,0 +1,187 @@
+use std::{fmt, io::Write, path::PathBuf, sync::Arc};
+use tracing::{field, span, Id, Level, Metadata, Subscriber};
+use tracing_appender::non_blocking::{NonBlocking, WorkerGuard};
+use tracing_subscriber::{
+    field::RecordFields,
+    filter::{Directive, FilterFn, Filtered},
+    fmt::{format, FormatFields, FormattedFields, MakeWriter},
+    layer::{Context, Layer},
+    registry::LookupSpan,
+};
+
+const ENV_ACCESS_LOG: &str = "LINKERD2_PROXY_ACCESS_LOG";
```

Comment:
Stepping back from the code for a second: how is this actually supposed to be used in practice? How would a Linkerd user configure the proxy so that it had a volume to which access log files could be written?

I can understand emitting logs to stdout or stderr, but it's not clear to me how output to a file could actually be used.

Reply:
Honestly, I have no idea; this is just what @tustvold's PR #601 did, and since that was opened by a potential user of this feature, I had kind of assumed they already knew how the logs were going to be collected, and that emitting the logs to a user configured file was the best way to integrate with whatever existing systems users are already using...

But, thinking through it...this would probably require adding some kind of volume mount to the proxy container, and because the proxy containers are added to pods by the proxy-injector _after_ the user applies the configuration for a pod, this might be kind of hard to do? Perhaps it's worth getting additional user input about how people want to get access logs out of Linkerd...

Reply:
I'm afraid my memory is a little bit hazy as to why I implemented it this way, but I think it was simply a case of replicating what Envoy does, combined with it being a simple way for the user to choose to be able to choose which of stdout or stderr to route logs to... There might have also been something about it being easier to write tests for, but I don't really recall I'm afraid.

For the reasons you articulate logging to a file in an environment making using of the admission webhook to inject the sidecars would be non-trivial, however, it could theoretically be used when doing injection manually. That being said I'm not aware of any environments doing this.

TLDR I don't think file output is needed per se, but I'm not sure it particularly hurts either?

Reply:
> combined with it being a simple way for the user to choose to be able to choose which of stdout or stderr to route logs to.

Ah, if the intention is that the user would just set the environment variable to `/dev/stderr` or `/dev/stdout` as appropriate, that _should_ work. I guess I'd have to revise the follow-up PR #1350 to add a separate env variable for controlling the access log format, though --- if the user is going to set the env var to `/dev/stderr`, we obviously can't enable JSON access logging with a filename extension in that case! :)